cip/1.accepted/CIP2017-06-18-multiple-graphs.adoc

```diff
@@ -0,0 +1,944 @@
+
+For the multiple property graphs execution model, it is necessary to define how query combinators combine the query contexts of all child queries into a new result query context (c.f. query composition).
+
+This CIP proposes that the multi-arm query combinators `UNION` (and `UNION ALL` respectively) combine their contexts according to the following rules:
+
+* Tabular data is combined as today, i.e. the tabular result is either a concatenation (`UNION ALL` case) or a distinct union (`UNION` case) of the tabular data from both arms
+*  All graphs from both arms are returned; if both arms return a graph with the same name, then the union of those graphs is returned under that shared name
+* If both graphs have specified a graph with the same name as their current source graph, then the union of those source graphs under that name again becomes the source graph for further processing.
+Otherwise, the default graph becomes the source graph for further processing.
+* If both graphs have specified a graph with the same name as their current target graph, then the union of those target graphs under that name again becomes the target graph for further processing.
+Otherwise, the default graph becomes the target graph for further processing.
+
+=== Expression evaluation
+
+Expressions are generally evaluated using the source graph of the current query context.
+
+=== Updating queries
+
+This CIP proposes the following update semantics for Cypher with support for multiple graphs:
+
+* All updating clauses read from the source graph and write to the target graph of their current query context.
+  More concretely:
+  ** Entities are always created in, updated in, and deleted from the currently provided target graph.
+  ** All entities of bound pattern variables in `CREATE` and `MERGE` are always added to the provided target graph of the current query context.
+  ** `MERGE` still uses the source graph to find existing entities
+  ** Deleting an entity only affects the provided target graph of the current query context.
+* Updating queries always return all variables and graphs in scope, i.e. the behave as if they would end in `RETURN *, GRAPHS *` (This syntax is introduced below).
+* Semantically, all effects of an updating clause must be made visible before proceeding with the execution of the next clause.
+In other words, a conforming implementation must ensure that a later clause always sees the complete set of updates of a preceding updating clause.
+
+This CIP proposes allowing `MERGE` to be followed by a non-empty, comma-separated list of bound variables for explicitly adding an entity to the target graph.
+
+=== Simulating the single graph execution model
+
+Execution under the single graph execution model can be simulated in the multiple graphs execution model by executing the query in an session context that uses the single graph as the default graph, and by running it on an empty initial query context.
```

Comment:
What about queries that end with an updating clause? They will return vs not return in the two models, no?